File: src/ehs/ACC.py
from m5.objects.EnergyCompressor import EnergyCompressor
from m5.params import *


class ACC(EnergyCompressor):
    """Adaptive Cache Compression (Alameldeen & Wood, ISCA 2004).

    One saturating counter of net cycles saved by compression gates
    the wrapped compressor. Fills compress while the counter is positive
    (Compression Mode) and store uncompressed while it is not (Regular Mode).
    A gated fill never runs the compressor, so Regular Mode also stops
    paying compression energy. The counter is fed by an
    ACCCache, which classifies hits.

    The GCP is architectural state (it survives power failures), so its size
    belongs in the controller's ckpt_reg_bytes.
    """

    type = "ACC"
    cxx_class = "gem5::compression::ACC"
    cxx_header = "ehs/acc.hh"

    # 16-bit signed counter
    gcp_min = Param.Int(-32768, "GCP saturation floor")
    gcp_max = Param.Int(32767, "GCP saturation ceiling")

    # Start in Compression Mode with modest confidence.
    # high enough that a few early unlucky decompressions do not
    # flip the mode before any evidence accumulates, low enough
    # that a genuinely useless compressor is switched off early in the run.
    gcp_init = Param.Int(1024, "Initial GCP value")

    # In Regular Mode nothing coallocates, so no reward could ever arriv.
    # Compressing every Nth fill anyway keeps evidence flowing at 1/N of
    # the energy. 0 disables sampling (Regular Mode then becomes permanent).
    sample_interval = Param.Unsigned(64,
        "Regular Mode fills between forced compressions (0 = never)")

    # The GCP is fed only by hits on blocks that compressed. A co-allocated hit
    # rewards it, a hit on a compressed block that earned nothing debits the
    # decompression latency. A compression that FAILS produces a full-size
    # block, which can never co-allocate and is never marked compressed, so it
    # generates neither signal for the rest of its life. Also it costs
    # latency and energy and tells the predictor nothing.
    #
    # Charging a failed attempt closes this loop using information the
    # compressor already produced. 0 disables it, which is the default: every
    # result taken before this parameter existed must reproduce byte for byte.
    failed_penalty = Param.Unsigned(0,
        "GCP cycles debited when a compression attempt comes back full size "
        "(0 = disabled, the published behaviour)")

File: src/ehs/acc.cc
#include "ehs/acc.hh"

#include <algorithm>

#include "base/logging.hh"
#include "base/trace.hh"
#include "debug/ACC.hh"
#include "params/ACC.hh"

namespace gem5
{
namespace compression
{

ACC::ACC(const Params &p)
  : EnergyCompressor(p),
    gcp(p.gcp_init),
    gcpMin(p.gcp_min),
    gcpMax(p.gcp_max),
    sampleInterval(p.sample_interval),
    gatedSinceSample(0),
    failedPenalty(p.failed_penalty),
    accStats(this)
{
    fatal_if(gcpMin >= gcpMax, "GCP saturation floor must be below ceiling");
    fatal_if(gcp < gcpMin || gcp > gcpMax,
             "Initial GCP outside its saturation range");
}

std::unique_ptr<Base::CompressionData>
ACC::runAndScore(const std::vector<Chunk>& chunks, Cycles& comp_lat,
                 Cycles& decomp_lat)
{
    auto comp_data = runCompressor(chunks, comp_lat, decomp_lat);

    // A full size result never coallocates and is never marked compressed,
    // so it generates neither a reward nor a penalty for the rest of its
    // life. The GCP only ever sees attempts that succeeded. The original
    // method does not need a penalty here because it runs the compressor
    // on every allocation whether or not it stores the result. Compression
    // cost is a constant there, and the computed size still feeds its
    // avoidable-miss test, so incompressible data simply stops earning
    // rewards.

    if (failedPenalty && comp_data->getSizeBits() >= blkSize * 8) {
        accStats.failedPenalties++;
        penalize(failedPenalty);
    }

    return comp_data;
}

std::unique_ptr<Base::CompressionData>
ACC::compress(const std::vector<Chunk>& chunks, Cycles& comp_lat,
              Cycles& decomp_lat)
{
    // The software compressibility hint outranks every policy below it
    noteHintAddr();
    if (hintSaysSkip()) {
        energyStats.hintSkips++;
        return passThrough(chunks, comp_lat, decomp_lat);
    }

    // Kagura's end-of-cycle override outranks everything below,
    if (kaguraRegular) {
        if (!dirtyOverride()) {
            accStats.kaguraGatedCompressions++;
            return passThrough(chunks, comp_lat, decomp_lat);
        }
        energyStats.dirtyOverrides++;
    }

    if (compressionEnabled()) {
        return runAndScore(chunks, comp_lat, decomp_lat);
    }

    // Regular Mode. Compress once every sample_interval fill anyway: a gated
    // fill can never co-allocate, so without sampling no reward could ever
    // arrive and the predictor would wedge at its floor.
    if (sampleInterval && ++gatedSinceSample >= sampleInterval) {
        gatedSinceSample = 0;
        accStats.sampledCompressions++;
        return runAndScore(chunks, comp_lat, decomp_lat);
    }

    accStats.gatedCompressions++;
    return passThrough(chunks, comp_lat, decomp_lat);
}

void
ACC::reward(int64_t cycles)
{
    const bool was_enabled = compressionEnabled();

    gcp = std::min(gcp + cycles, gcpMax);
    accStats.rewards++;

    if (compressionEnabled() != was_enabled) {
        accStats.modeSwitches++;
        DPRINTF(ACC, "GCP=%lld: entering Compression Mode\n", gcp);
    }
}

void
ACC::penalize(int64_t cycles)
{
    const bool was_enabled = compressionEnabled();

    gcp = std::max(gcp - cycles, gcpMin);
    accStats.penalties++;

    if (compressionEnabled() != was_enabled) {
        accStats.modeSwitches++;
        DPRINTF(ACC, "GCP=%lld: entering Regular Mode\n", gcp);
    }
}

ACC::ACCStats::ACCStats(statistics::Group *parent)
  : statistics::Group(parent),
    ADD_STAT(rewards, statistics::units::Count::get(),
             "GCP credits: hits attributed to compression's extra capacity"),
    ADD_STAT(penalties, statistics::units::Count::get(),
             "GCP debits: decompression latency paid for nothing"),
    ADD_STAT(gatedCompressions, statistics::units::Count::get(),
             "Fills stored uncompressed because the GCP said not to bother"),
    ADD_STAT(sampledCompressions, statistics::units::Count::get(),
             "Regular Mode fills compressed anyway to keep evidence flowing"),
    ADD_STAT(modeSwitches, statistics::units::Count::get(),
             "Compression Mode <-> Regular Mode transitions"),
    ADD_STAT(kaguraGatedCompressions, statistics::units::Count::get(),
             "Fills stored uncompressed on Kagura's end-of-cycle override"),
    ADD_STAT(failedPenalties, statistics::units::Count::get(),
             "Compression attempts that came back full size and were charged "
             "to the GCP (0 unless failed_penalty is set)")
{
}

} // namespace compression
} // namespace gem5
